### PR TITLE
feat: add support for getting and setting table IAM policy

### PR DESCRIPTION
**Pull request by steffnay**
Adds support for interacting with IAM policies for BigQuery table resources.

- [x] Ensure the tests and linter pass
- [x] Code coverage does not decrease (if any source code was changed)
- [x] Appropriate docs were updated (if necessary)

🦕

### CODE REVIEW

**Comment by shollyman** [comment]
Testing is picking up unit test problems with pyarrow tests.  Looks like they just jumped from 0.17.1 to 1.0, so possibly some more shenanigans afoot with changes there.

**Comment by shollyman** [comment]
Given it's friday afternoon now, punting a release window on this until next week.

**Comment by shollyman** [comment]
And ignore the previous two comments, I've been typing in the wrong PR

File: google/cloud/bigquery/client.py [hunk 1]
```diff
@@ -605,6 +606,70 @@ def get_dataset(self, dataset_ref, retry=DEFAULT_RETRY, timeout=None):
...
+        if not isinstance(table, (Table, TableReference)):
+            raise TypeError("table must be a Table or TableReference")
+
+        if requested_policy_version != 1:
+            raise ValueError("only IAM policy version 1 is supported")
+
+        options = {}
+        body = {}
+        options["requestedPolicyVersion"] = requested_policy_version
+        body["options"] = options
```
Comment (tseaver) [inline google/cloud/bigquery/client.py:621]: Simpler / clearer as a pure dict literal., e.g.:
```suggestion
        body = {"options": {"requestedPolicyVersion": 1}}
```

File: google/cloud/bigquery/client.py [hunk 2]
```diff
@@ -605,6 +606,70 @@ def get_dataset(self, dataset_ref, retry=DEFAULT_RETRY, timeout=None):
...
+
+        path = "%s:getIamPolicy" % (table.path)
```
Comment (tseaver) [inline google/cloud/bigquery/client.py:623]: Prefer `"".format"` over the `%` operator, e.g.:
```suggestion
        path = "{}:getIamPolicy".format(table.path)

File: google/cloud/bigquery/client.py [hunk 3]
```diff
@@ -605,6 +606,70 @@ def get_dataset(self, dataset_ref, retry=DEFAULT_RETRY, timeout=None):
...
+    def set_iam_policy(
+        self, table, policy, updateMask=None, retry=DEFAULT_RETRY, timeout=None,
+    ):
+        if not isinstance(table, (Table, TableReference)):
+            raise TypeError("table must be a Table or TableReference")
+
+        if not isinstance(policy, (Policy)):
+            raise TypeError("policy must be a Policy")
+
+        body = {}
```
Comment (tseaver) [inline google/cloud/bigquery/client.py:640]: Likewise, simplify here:
```suggestion
        body = {"policy": policy.to_api_repr()}
```

File: google/cloud/bigquery/client.py [hunk 4]
```diff
@@ -605,6 +606,70 @@ def get_dataset(self, dataset_ref, retry=DEFAULT_RETRY, timeout=None):
...
+
+        if updateMask is not None:
+            body["updateMask"] = updateMask
+
+        body["policy"] = policy.to_api_repr()
+
+        path = "%s:setIamPolicy" % (table.path)
```
Comment (tseaver) [inline google/cloud/bigquery/client.py:647]: Again here.:
```suggestion
        path = "{}:getIamPolicy".format(table.path)

File: google/cloud/bigquery/client.py [hunk 5]
```diff
@@ -605,6 +606,70 @@ def get_dataset(self, dataset_ref, retry=DEFAULT_RETRY, timeout=None):
...
+
+    def test_iam_permissions(
+        self, table, permissions, retry=DEFAULT_RETRY, timeout=None,
+    ):
+        if not isinstance(table, (Table, TableReference)):
+            raise TypeError("table must be a Table or TableReference")
+
+        body = {}
+
+        body["permissions"] = permissions
```
Comment (tseaver) [inline google/cloud/bigquery/client.py:663]: Simplify again:
```suggestion
        body = {"permissions": permissions}
```

File: google/cloud/bigquery/client.py [hunk 6]
```diff
@@ -605,6 +606,70 @@ def get_dataset(self, dataset_ref, retry=DEFAULT_RETRY, timeout=None):
...
+
+        path = "%s:testIamPermissions" % (table.path)
```
Comment (tseaver) [inline google/cloud/bigquery/client.py:665]: And here:.:
```suggestion
        path = "{}:getIamPolicy".format(table.path)

File: tests/system.py [hunk 7]
```diff
@@ -1407,6 +1408,54 @@ def test_copy_table(self):
         got_rows = self._fetch_single_page(dest_table)
         self.assertTrue(len(got_rows) > 0)
 
+    def test_iam_policy(self):
```
Comment (tseaver) [inline tests/system.py:1411]:
```suggestion
    def test_get_set_iam_policy(self):
```

File: tests/unit/test_client.py [hunk 8]
```diff
@@ -1748,6 +1748,220 @@ def test_get_table_sets_user_agent(self):
...
+        from google.cloud.bigquery.iam import BIGQUERY_DATA_OWNER_ROLE
+        from google.cloud.bigquery.iam import BIGQUERY_DATA_EDITOR_ROLE
+        from google.cloud.bigquery.iam import BIGQUERY_DATA_VIEWER_ROLE
+        from google.api_core.iam import Policy
+
+        PATH = "/projects/%s/datasets/%s/tables/%s:getIamPolicy" % (
+            self.PROJECT,
+            self.DS_ID,
+            self.TABLE_ID,
+        )
```
Comment (tseaver) [inline tests/unit/test_client.py:1761]: Again, prefer `str.format`:

```python
        PATH = "/projects/{}/datasets/{}/tables/{}:getIamPolicy".format(
            self.PROJECT,
            self.DS_ID,
            self.TABLE_ID,
        )
```

File: tests/unit/test_client.py [hunk 9]
```diff
@@ -1748,6 +1748,220 @@ def test_get_table_sets_user_agent(self):
...
+        self.assertEqual(policy.etag, RETURNED["etag"])
+        self.assertEqual(policy.version, RETURNED["version"])
+        self.assertEqual(dict(policy), EXPECTED)
+
+    def test_get_iam_policy_w_invalid_table(self):
+        creds = _make_credentials()
+        http = object()
+        client = self._make_one(project=self.PROJECT, credentials=creds, _http=http)
+
+        table_resource_string = "projects/%s/datasets/%s/tables/%s" % (
```
Comment (tseaver) [inline tests/unit/test_client.py:1806]: Likewise here.

**Comment by brandonjchoi** [comment]
When will this be released @tseaver @shollyman ?

**Pull request merged**